File: scripts/learning/semantic_action_store.py
#!/usr/bin/env python3
"""SQLite-backed semantic action store and migration helpers."""

from __future__ import annotations

import json
import sqlite3
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable


SCHEMA_VERSION = "openstaff.semantic-action.v0"
MIGRATIONS_DIR = Path(__file__).resolve().parent / "migrations" / "semantic_actions"


@dataclass(frozen=True)
class Migration:
    version: int
    name: str
    up_path: Path
    down_path: Path


MIGRATIONS = [
    Migration(
        version=1,
        name="semantic_actions",
        up_path=MIGRATIONS_DIR / "0001_semantic_actions.up.sql",
        down_path=MIGRATIONS_DIR / "0001_semantic_actions.down.sql",
    )
]


@dataclass(frozen=True)
class SemanticActionRecord:
    action_id: str
    session_id: str
    action_type: str
    selector: dict[str, Any]
    created_at: str
    updated_at: str
    args: dict[str, Any] = field(default_factory=dict)
    context: dict[str, Any] = field(default_factory=dict)
    confidence: float = 0.0
    source_event_ids: list[str] = field(default_factory=list)
    source_frame_ids: list[str] = field(default_factory=list)
    schema_version: str = SCHEMA_VERSION
    task_id: str | None = None
    turn_id: str | None = None
    trace_id: str | None = None
    step_id: str | None = None
    step_index: int | None = None
    source_path: str | None = None
    preferred_locator_type: str | None = None
    manual_review_required: bool = False
    legacy_coordinate: dict[str, Any] | None = None


@dataclass(frozen=True)
class SemanticActionTargetRecord:
    target_id: str
    target_role: str
    ordinal: int
    selector: dict[str, Any]
    created_at: str
    context: dict[str, Any] = field(default_factory=dict)
    locator_type: str | None = None
    confidence: float | None = None
    is_preferred: bool = False


@dataclass(frozen=True)
class SemanticActionAssertionRecord:
    assertion_id: str
    assertion_type: str
    assertion: dict[str, Any]
    created_at: str
    source: str
    is_required: bool = True
    ordinal: int = 0


@dataclass(frozen=True)
class SemanticActionExecutionLogRecord:
    execution_log_id: str
    status: str
    executed_at: str
    selector_hit_path: list[str] = field(default_factory=list)
    result: dict[str, Any] = field(default_factory=dict)
    trace_id: str | None = None
    component: str | None = None
    error_code: str | None = None
    duration_ms: int | None = None
    execution_log_path: str | None = None
    execution_result_path: str | None = None
    review_id: str | None = None


def _json_dumps(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, sort_keys=True)


def _json_loads(value: str | None, default: Any) -> Any:
    if value is None or value == "":
        return default
    return json.loads(value)


def connect_sqlite(db_path: Path) -> sqlite3.Connection:
    db_path = Path(db_path)
    db_path.parent.mkdir(parents=True, exist_ok=True)
    connection = sqlite3.connect(db_path)
    connection.row_factory = sqlite3.Row
    connection.execute("PRAGMA foreign_keys=ON")
    return connection


class SemanticActionMigrationManager:
    def __init__(self, db_path: Path) -> None:
        self.db_path = Path(db_path)

    def applied_versions(self) -> list[int]:
        with connect_sqlite(self.db_path) as connection:
            self._ensure_migration_table(connection)
            rows = connection.execute(
                "SELECT version FROM _openstaff_schema_migrations ORDER BY version ASC"
            ).fetchall()
        return [int(row["version"]) for row in rows]

    def migrate_up(self) -> list[int]:
        applied: list[int] = []
        with connect_sqlite(self.db_path) as connection:
            self._ensure_migration_table(connection)
            existing = set(self._fetch_existing_versions(connection))
            for migration in MIGRATIONS:
                if migration.version in existing:
                    continue
                connection.executescript(migration.up_path.read_text(encoding="utf-8"))
                connection.execute(
                    """
                    INSERT INTO _openstaff_schema_migrations (version, name, applied_at)
                    VALUES (?, ?, strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))
                    """,
                    (migration.version, migration.name),
                )
                applied.append(migration.version)
        return applied

    def migrate_down(self) -> list[int]:
        rolled_back: list[int] = []
        with connect_sqlite(self.db_path) as connection:
            self._ensure_migration_table(connection)
            existing = set(self._fetch_existing_versions(connection))
            for migration in reversed(MIGRATIONS):
                if migration.version not in existing:
                    continue
                connection.executescript(migration.down_path.read_text(encoding="utf-8"))
                connection.execute(
                    "DELETE FROM _openstaff_schema_migrations WHERE version = ?",
                    (migration.version,),
                )
                rolled_back.append(migration.version)
        return rolled_back

    @staticmethod
    def _ensure_migration_table(connection: sqlite3.Connection) -> None:
        connection.execute(
            """
            CREATE TABLE IF NOT EXISTS _openstaff_schema_migrations (
                version INTEGER PRIMARY KEY,
                name TEXT NOT NULL,
                applied_at TEXT NOT NULL
            )
            """
        )

    @staticmethod
    def _fetch_existing_versions(connection: sqlite3.Connection) -> list[int]:
        rows = connection.execute(
            "SELECT version FROM _openstaff_schema_migrations ORDER BY version ASC"
        ).fetchall()
        return [int(row["version"]) for row in rows]


class SemanticActionRepository:
    def __init__(self, db_path: Path) -> None:
        self.db_path = Path(db_path)

    def clear_all(self) -> None:
        with connect_sqlite(self.db_path) as connection:
            connection.execute("DELETE FROM action_execution_logs")
            connection.execute("DELETE FROM action_assertions")
            connection.execute("DELETE FROM action_targets")
            connection.execute("DELETE FROM semantic_actions")

    def count_actions(self) -> int:
        with connect_sqlite(self.db_path) as connection:
            row = connection.execute("SELECT COUNT(*) AS count FROM semantic_actions").fetchone()
        return int(row["count"]) if row else 0

    def replace_action(
        self,
        action: SemanticActionRecord,
        *,
        targets: Iterable[SemanticActionTargetRecord] = (),
        assertions: Iterable[SemanticActionAssertionRecord] = (),
        execution_logs: Iterable[SemanticActionExecutionLogRecord] = (),
    ) -> None:
        target_records = list(targets)
        assertion_records = list(assertions)
        execution_log_records = list(execution_logs)

        with connect_sqlite(self.db_path) as connection:
            connection.execute(
                """
                INSERT INTO semantic_actions (
                    action_id,
                    schema_version,
                    session_id,
                    task_id,
                    turn_id,
                    trace_id,
                    step_id,
                    step_index,
                    action_type,
                    selector_json,
                    args_json,
                    context_json,
                    confidence,
                    source_event_ids,
                    source_frame_ids,
                    source_path,
                    preferred_locator_type,
                    manual_review_required,
                    legacy_coordinate_json,
                    created_at,
                    updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(action_id) DO UPDATE SET
                    schema_version = excluded.schema_version,
                    session_id = excluded.session_id,
                    task_id = excluded.task_id,
                    turn_id = excluded.turn_id,
                    trace_id = excluded.trace_id,
                    step_id = excluded.step_id,
                    step_index = excluded.step_index,
                    action_type = excluded.action_type,
                    selector_json = excluded.selector_json,
                    args_json = excluded.args_json,
                    context_json = excluded.context_json,
                    confidence = excluded.confidence,
                    source_event_ids = excluded.source_event_ids,
                    source_frame_ids = excluded.source_frame_ids,
                    source_path = excluded.source_path,
                    preferred_locator_type = excluded.preferred_locator_type,
                    manual_review_required = excluded.manual_review_required,
                    legacy_coordinate_json = excluded.legacy_coordinate_json,
                    created_at = excluded.created_at,
                    updated_at = excluded.updated_at
                """,
                (
                    action.action_id,
                    action.schema_version,
                    action.session_id,
                    action.task_id,
                    action.turn_id,
                    action.trace_id,
                    action.step_id,
                    action.step_index,
                    action.action_type,
                    _json_dumps(action.selector),
                    _json_dumps(action.args),
                    _json_dumps(action.context),
                    action.confidence,
                    _json_dumps(action.source_event_ids),
                    _json_dumps(action.source_frame_ids),
                    action.source_path,
                    action.preferred_locator_type,
                    1 if action.manual_review_required else 0,
                    _json_dumps(action.legacy_coordinate) if action.legacy_coordinate is not None else None,
                    action.created_at,
                    action.updated_at,
                ),
            )

            for table_name in ("action_targets", "action_assertions", "action_execution_logs"):
                connection.execute(f"DELETE FROM {table_name} WHERE action_id = ?", (action.action_id,))

            if target_records:
                connection.executemany(
                    """
                    INSERT INTO action_targets (
                        target_id,
                        action_id,
                        target_role,
                        ordinal,
                        locator_type,
                        selector_json,
                        context_json,
                        confidence,
                        is_preferred,
                        created_at
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    [
                        (
                            target.target_id,
                            action.action_id,
                            target.target_role,
                            target.ordinal,
                            target.locator_type,
                            _json_dumps(target.selector),
                            _json_dumps(target.context),
                            target.confidence,
                            1 if target.is_preferred else 0,
                            target.created_at,
                        )
                        for target in target_records
                    ],
                )

            if assertion_records:
                connection.executemany(
                    """
                    INSERT INTO action_assertions (
                        assertion_id,
                        action_id,
                        assertion_type,
                        assertion_json,
                        is_required,
                        ordinal,
                        source,
                        created_at
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    [
                        (
                            assertion.assertion_id,
                            action.action_id,
                            assertion.assertion_type,
                            _json_dumps(assertion.assertion),
                            1 if assertion.is_required else 0,
                            assertion.ordinal,
                            assertion.source,
                            assertion.created_at,
                        )
                        for assertion in assertion_records
                    ],
                )

            if execution_log_records:
                connection.executemany(
                    """
                    INSERT INTO action_execution_logs (
                        execution_log_id,
                        action_id,
                        trace_id,
                        component,
                        status,
                        error_code,
                        selector_hit_path_json,
                        result_json,
                        duration_ms,
                        execution_log_path,
                        execution_result_path,
                        review_id,
                        executed_at
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    [
                        (
                            execution_log.execution_log_id,
                            action.action_id,
                            execution_log.trace_id,
                            execution_log.component,
                            execution_log.status,
                            execution_log.error_code,
                            _json_dumps(execution_log.selector_hit_path),
                            _json_dumps(execution_log.result),
                            execution_log.duration_ms,
                            execution_log.execution_log_path,
                            execution_log.execution_result_path,
                            execution_log.review_id,
                            execution_log.executed_at,
                        )
                        for execution_log in execution_log_records
                    ],
                )

    def get_action(self, action_id: str) -> dict[str, Any] | None:
        with connect_sqlite(self.db_path) as connection:
            row = connection.execute(
                "SELECT * FROM semantic_actions WHERE action_id = ?",
                (action_id,),
            ).fetchone()
            if row is None:
                return None
            return self._hydrate_action(connection, row)

    def list_actions(
        self,
        *,
        session_id: str | None = None,
        action_type: str | None = None,
        limit: int | None = None,
    ) -> list[dict[str, Any]]:
        clauses: list[str] = []
        params: list[Any] = []

        if session_id:
            clauses.append("session_id = ?")
            params.append(session_id)
        if action_type:
            clauses.append("action_type = ?")
            params.append(action_type)

        query = "SELECT * FROM semantic_actions"
        if clauses:
            query += " WHERE " + " AND ".join(clauses)
        query += " ORDER BY created_at ASC, action_id ASC"
        if limit is not None:
            query += " LIMIT ?"
            params.append(limit)

        with connect_sqlite(self.db_path) as connection:
            rows = connection.execute(query, tuple(params)).fetchall()
            return [self._hydrate_action(connection, row) for row in rows]

    def _hydrate_action(self, connection: sqlite3.Connection, row: sqlite3.Row) -> dict[str, Any]:
        payload = dict(row)
        payload["selector_json"] = _json_loads(payload.get("selector_json"), {})
        payload["args_json"] = _json_loads(payload.get("args_json"), {})
        payload["context_json"] = _json_loads(payload.get("context_json"), {})
        payload["source_event_ids"] = _json_loads(payload.get("source_event_ids"), [])
        payload["source_frame_ids"] = _json_loads(payload.get("source_frame_ids"), [])
        payload["legacy_coordinate_json"] = _json_loads(payload.get("legacy_coordinate_json"), None)
        payload["manual_review_required"] = bool(payload.get("manual_review_required"))
        payload["targets"] = self._load_targets(connection, payload["action_id"])
        payload["assertions"] = self._load_assertions(connection, payload["action_id"])
        payload["execution_logs"] = self._load_execution_logs(connection, payload["action_id"])
        return payload

    def _load_targets(self, connection: sqlite3.Connection, action_id: str) -> list[dict[str, Any]]:
        rows = connection.execute(
            """
            SELECT *
            FROM action_targets
            WHERE action_id = ?
            ORDER BY ordinal ASC, target_id ASC
            """,
            (action_id,),
        ).fetchall()
        targets: list[dict[str, Any]] = []
        for row in rows:
            payload = dict(row)
            payload["selector_json"] = _json_loads(payload.get("selector_json"), {})
            payload["context_json"] = _json_loads(payload.get("context_json"), {})
            payload["is_preferred"] = bool(payload.get("is_preferred"))
            targets.append(payload)
        return targets

    def _load_assertions(self, connection: sqlite3.Connection, action_id: str) -> list[dict[str, Any]]:
        rows = connection.execute(
            """
            SELECT *
            FROM action_assertions
            WHERE action_id = ?
            ORDER BY ordinal ASC, assertion_id ASC
            """,
            (action_id,),
        ).fetchall()
        assertions: list[dict[str, Any]] = []
        for row in rows:
            payload = dict(row)
            payload["assertion_json"] = _json_loads(payload.get("assertion_json"), {})
            payload["is_required"] = bool(payload.get("is_required"))
            assertions.append(payload)
        return assertions

    def _load_execution_logs(self, connection: sqlite3.Connection, action_id: str) -> list[dict[str, Any]]:
        rows = connection.execute(
            """
            SELECT *
            FROM action_execution_logs
            WHERE action_id = ?
            ORDER BY executed_at ASC, execution_log_id ASC
            """,
            (action_id,),
        ).fetchall()
        execution_logs: list[dict[str, Any]] = []
        for row in rows:
            payload = dict(row)
            payload["selector_hit_path_json"] = _json_loads(payload.get("selector_hit_path_json"), [])
            payload["result_json"] = _json_loads(payload.get("result_json"), {})
            execution_logs.append(payload)
        return execution_logs

File: apps/macos/Sources/OpenStaffReplayVerifyCLI/SemanticActionStore.swift
import Foundation
import SQLite3

typealias SemanticJSONObject = [String: Any]

enum SemanticActionStoreError: LocalizedError {
    case openDatabase(String)
    case prepareStatement(String)
    case queryFailed(String)
    case invalidJSON(String)

    var errorDescription: String? {
        switch self {
        case .openDatabase(let message):
            return "Semantic action store open failed: \(message)"
        case .prepareStatement(let message):
            return "Semantic action store prepare failed: \(message)"
        case .queryFailed(let message):
            return "Semantic action store query failed: \(message)"
        case .invalidJSON(let message):
            return "Semantic action store JSON decode failed: \(message)"
        }
    }
}

struct SemanticActionStoreTargetRecord {
    let targetId: String
    let targetRole: String
    let ordinal: Int
    let locatorType: String?
    let selector: SemanticJSONObject
    let isPreferred: Bool
}

struct SemanticActionStoreAssertionRecord {
    let assertionId: String
    let assertionType: String
    let payload: SemanticJSONObject
    let isRequired: Bool
    let ordinal: Int
}

struct SemanticActionStoreAction {
    let actionId: String
    let sessionId: String
    let taskId: String?
    let traceId: String?
    let stepId: String?
    let actionType: String
    let selector: SemanticJSONObject
    let args: SemanticJSONObject
    let context: SemanticJSONObject
    let preferredLocatorType: String?
    let manualReviewRequired: Bool
    let createdAt: String
    let updatedAt: String
    let targets: [SemanticActionStoreTargetRecord]
    let assertions: [SemanticActionStoreAssertionRecord]
}

struct SemanticActionStoreExecutionLogRecord {
    let executionLogId: String
    let actionId: String
    let traceId: String?
    let component: String
    let status: String
    let errorCode: String?
    let selectorHitPath: [String]
    let result: SemanticJSONObject
    let durationMs: Int
    let executedAt: String
}

final class SemanticActionSQLiteStore {
    private let databaseURL: URL

    init(databaseURL: URL) {
        self.databaseURL = databaseURL
    }

    func fetchAction(actionId: String) throws -> SemanticActionStoreAction? {
        try withDatabase { database in
            let statement = try prepare(
                database,
                sql: """
                SELECT
                    action_id,
                    session_id,
                    task_id,
                    trace_id,
                    step_id,
                    action_type,
                    selector_json,
                    args_json,
                    context_json,
                    preferred_locator_type,
                    manual_review_required,
                    created_at,
                    updated_at
                FROM semantic_actions
                WHERE action_id = ?
                LIMIT 1
                """
            )
            defer { sqlite3_finalize(statement) }

            bindText(actionId, to: statement, index: 1)
            let result = sqlite3_step(statement)
            guard result == SQLITE_ROW else {
                if result == SQLITE_DONE {
                    return nil
                }
                throw SemanticActionStoreError.queryFailed(lastErrorMessage(database))
            }

            let fetchedActionId = text(statement, index: 0) ?? actionId
            return SemanticActionStoreAction(
                actionId: fetchedActionId,
                sessionId: text(statement, index: 1) ?? "",
                taskId: text(statement, index: 2),
                traceId: text(statement, index: 3),
                stepId: text(statement, index: 4),
                actionType: text(statement, index: 5) ?? "",
                selector: try object(statement, index: 6, label: "selector_json"),
                args: try object(statement, index: 7, label: "args_json"),
                context: try object(statement, index: 8, label: "context_json"),
                preferredLocatorType: text(statement, index: 9),
                manualReviewRequired: integer(statement, index: 10) != 0,
                createdAt: text(statement, index: 11) ?? "",
                updatedAt: text(statement, index: 12) ?? "",
                targets: try fetchTargets(database: database, actionId: fetchedActionId),
                assertions: try fetchAssertions(database: database, actionId: fetchedActionId)
            )
        }
    }

    func appendExecutionLog(_ log: SemanticActionStoreExecutionLogRecord) throws {
        try withDatabase { database in
            let statement = try prepare(
                database,
                sql: """
                INSERT INTO action_execution_logs (
                    execution_log_id,
                    action_id,
                    trace_id,
                    component,
                    status,
                    error_code,
                    selector_hit_path_json,
                    result_json,
                    duration_ms,
                    execution_log_path,
                    execution_result_path,
                    review_id,
                    executed_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, NULL, NULL, NULL, ?)
                """
            )
            defer { sqlite3_finalize(statement) }

            bindText(log.executionLogId, to: statement, index: 1)
            bindText(log.actionId, to: statement, index: 2)
            bindOptionalText(log.traceId, to: statement, index: 3)
            bindText(log.component, to: statement, index: 4)
            bindText(log.status, to: statement, index: 5)
            bindOptionalText(log.errorCode, to: statement, index: 6)
            bindText(jsonString(log.selectorHitPath), to: statement, index: 7)
            bindText(jsonString(log.result), to: statement, index: 8)
            sqlite3_bind_int64(statement, 9, sqlite3_int64(log.durationMs))
            bindText(log.executedAt, to: statement, index: 10)

            guard sqlite3_step(statement) == SQLITE_DONE else {
                throw SemanticActionStoreError.queryFailed(lastErrorMessage(database))
            }
        }
    }

    private func fetchTargets(
        database: OpaquePointer?,
        actionId: String
    ) throws -> [SemanticActionStoreTargetRecord] {
        let statement = try prepare(
            database,
            sql: """
            SELECT target_id, target_role, ordinal, locator_type, selector_json, is_preferred
            FROM action_targets
            WHERE action_id = ?
            ORDER BY ordinal ASC, target_id ASC
            """
        )
        defer { sqlite3_finalize(statement) }

        bindText(actionId, to: statement, index: 1)
        var rows: [SemanticActionStoreTargetRecord] = []
        while true {
            let result = sqlite3_step(statement)
            if result == SQLITE_DONE {
                break
            }
            guard result == SQLITE_ROW else {
                throw SemanticActionStoreError.queryFailed(lastErrorMessage(database))
            }
            rows.append(
                SemanticActionStoreTargetRecord(
                    targetId: text(statement, index: 0) ?? "",
                    targetRole: text(statement, index: 1) ?? "",
                    ordinal: Int(integer(statement, index: 2)),
                    locatorType: text(statement, index: 3),
                    selector: try object(statement, index: 4, label: "action_targets.selector_json"),
                    isPreferred: integer(statement, index: 5) != 0
                )
            )
        }
        return rows
    }

    private func fetchAssertions(
        database: OpaquePointer?,
        actionId: String
    ) throws -> [SemanticActionStoreAssertionRecord] {
        let statement = try prepare(
            database,
            sql: """
            SELECT assertion_id, assertion_type, assertion_json, is_required, ordinal
            FROM action_assertions
            WHERE action_id = ?
            ORDER BY ordinal ASC, assertion_id ASC
            """
        )
        defer { sqlite3_finalize(statement) }

        bindText(actionId, to: statement, index: 1)
        var rows: [SemanticActionStoreAssertionRecord] = []
        while true {
            let result = sqlite3_step(statement)
            if result == SQLITE_DONE {
                break
            }
            guard result == SQLITE_ROW else {
                throw SemanticActionStoreError.queryFailed(lastErrorMessage(database))
            }
            rows.append(
                SemanticActionStoreAssertionRecord(
                    assertionId: text(statement, index: 0) ?? "",
                    assertionType: text(statement, index: 1) ?? "",
                    payload: try object(statement, index: 2, label: "action_assertions.assertion_json"),
                    isRequired: integer(statement, index: 3) != 0,
                    ordinal: Int(integer(statement, index: 4))
                )
            )
        }
        return rows
    }

    private func withDatabase<T>(_ body: (OpaquePointer?) throws -> T) throws -> T {
        var database: OpaquePointer?
        guard sqlite3_open_v2(
            databaseURL.path,
            &database,
            SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX,
            nil
        ) == SQLITE_OK else {
            defer { sqlite3_close(database) }
            throw SemanticActionStoreError.openDatabase(lastErrorMessage(database))
        }
        defer { sqlite3_close(database) }
        return try body(database)
    }

    private func prepare(_ database: OpaquePointer?, sql: String) throws -> OpaquePointer? {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(database, sql, -1, &statement, nil) == SQLITE_OK else {
            throw SemanticActionStoreError.prepareStatement(lastErrorMessage(database))
        }
        return statement
    }

    private func text(_ statement: OpaquePointer?, index: Int32) -> String? {
        guard let pointer = sqlite3_column_text(statement, index) else {
            return nil
        }
        return String(cString: pointer)
    }

    private func integer(_ statement: OpaquePointer?, index: Int32) -> Int64 {
        sqlite3_column_int64(statement, index)
    }

    private func object(_ statement: OpaquePointer?, index: Int32, label: String) throws -> SemanticJSONObject {
        guard let raw = text(statement, index: index), !raw.isEmpty else {
            return [:]
        }
        guard let data = raw.data(using: .utf8),
              let value = try JSONSerialization.jsonObject(with: data) as? SemanticJSONObject else {
            throw SemanticActionStoreError.invalidJSON(label)
        }
        return value
    }

    private func bindText(_ value: String, to statement: OpaquePointer?, index: Int32) {
        sqlite3_bind_text(statement, index, value, -1, sqliteTransientDestructor)
    }

    private func bindOptionalText(_ value: String?, to statement: OpaquePointer?, index: Int32) {
        guard let value else {
            sqlite3_bind_null(statement, index)
            return
        }
        bindText(value, to: statement, index: index)
    }

    private func jsonString(_ value: Any) -> String {
        guard JSONSerialization.isValidJSONObject(value),
              let data = try? JSONSerialization.data(withJSONObject: value, options: [.sortedKeys]),
              let text = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return text
    }

    private func lastErrorMessage(_ database: OpaquePointer?) -> String {
        guard let database else {
            return "unknown sqlite error"
        }
        return String(cString: sqlite3_errmsg(database))
    }
}

private let sqliteTransientDestructor = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
